File: src/core/maze.ts
import { ALL_WALLS, EAST, NORTH, OPPOSITE, SOUTH, WEST, type Cell, type Direction } from './types'

export class Maze {
  readonly rows: number
  readonly cols: number
  /** walls[row][col] = bitmask of remaining walls */
  private walls: Uint8Array[]

  constructor(rows: number, cols: number) {
    this.rows = rows
    this.cols = cols
    // Every cell starts with all 4 walls
    this.walls = Array.from({ length: rows }, () => new Uint8Array(cols).fill(ALL_WALLS))
  }

  inBounds(row: number, col: number): boolean {
    return row >= 0 && row < this.rows && col >= 0 && col < this.cols
  }

  hasWall(cell: Cell, dir: Direction): boolean {
    return (this.walls[cell.row][cell.col] & dir) !== 0
  }

  /** Remove the wall between two adjacent cells */
  removeWall(a: Cell, b: Cell): void {
    const dr = b.row - a.row
    const dc = b.col - a.col
    let dir: Direction
    if (dr === -1) dir = NORTH
    else if (dr === 1) dir = SOUTH
    else if (dc === 1) dir = EAST
    else if (dc === -1) dir = WEST
    else throw new Error(`Cells are not adjacent: (${a.row},${a.col}) and (${b.row},${b.col})`)

    this.walls[a.row][a.col] &= ~dir
    this.walls[b.row][b.col] &= ~OPPOSITE[dir]
  }

  /** All valid neighbours of a cell (in bounds), regardless of walls */
  neighbors(cell: Cell): Cell[] {
    const { row, col } = cell
    const result: Cell[] = []
    if (row > 0)            result.push({ row: row - 1, col })
    if (row < this.rows - 1) result.push({ row: row + 1, col })
    if (col < this.cols - 1) result.push({ row, col: col + 1 })
    if (col > 0)            result.push({ row, col: col - 1 })
    return result
  }

  /** Neighbours reachable without crossing a wall */
  passableNeighbors(cell: Cell): Cell[] {
    const dirs: Direction[] = [NORTH, SOUTH, EAST, WEST]
    const deltas: [number, number][] = [[-1, 0], [1, 0], [0, 1], [0, -1]]
    const result: Cell[] = []
    for (let i = 0; i < 4; i++) {
      if (!this.hasWall(cell, dirs[i])) {
        const nr = cell.row + deltas[i][0]
        const nc = cell.col + deltas[i][1]
        if (this.inBounds(nr, nc)) result.push({ row: nr, col: nc })
      }
    }
    return result
  }

  cellKey(cell: Cell): number {
    return cell.row * this.cols + cell.col
  }

  fromKey(key: number): Cell {
    return { row: Math.floor(key / this.cols), col: key % this.cols }
  }
}

File: python/maze.py
"""
Maze — miroir exact de src/core/maze.ts
Bitmask : NORTH=1  SOUTH=2  EAST=4  WEST=8
"""
import random

NORTH = 1
SOUTH = 2
EAST  = 4
WEST  = 8
ALL_WALLS = NORTH | SOUTH | EAST | WEST

OPPOSITE = { NORTH: SOUTH, SOUTH: NORTH, EAST: WEST, WEST: EAST }
DELTA    = { NORTH: (-1, 0), SOUTH: (1, 0), EAST: (0, 1), WEST: (0, -1) }
DIRS     = [NORTH, SOUTH, EAST, WEST]


class Maze:
    def __init__(self, rows: int, cols: int):
        self.rows = rows
        self.cols = cols
        # walls[row][col] = bitmask des murs restants
        self.walls = [[ALL_WALLS] * cols for _ in range(rows)]

    def in_bounds(self, row: int, col: int) -> bool:
        return 0 <= row < self.rows and 0 <= col < self.cols

    def has_wall(self, row: int, col: int, direction: int) -> bool:
        return bool(self.walls[row][col] & direction)

    def remove_wall(self, r1: int, c1: int, r2: int, c2: int) -> None:
        dr, dc = r2 - r1, c2 - c1
        if   (dr, dc) == (-1, 0): d = NORTH
        elif (dr, dc) == ( 1, 0): d = SOUTH
        elif (dr, dc) == ( 0, 1): d = EAST
        else:                      d = WEST
        self.walls[r1][c1] &= ~d
        self.walls[r2][c2] &= ~OPPOSITE[d]

    def passable_neighbors(self, row: int, col: int):
        result = []
        for d in DIRS:
            if not self.has_wall(row, col, d):
                dr, dc = DELTA[d]
                nr, nc = row + dr, col + dc
                if self.in_bounds(nr, nc):
                    result.append((nr, nc))
        return result


def generate_maze(rows: int, cols: int, extra_openings: int = 0) -> Maze:
    """Recursive backtracker + extra openings aléatoires."""
    m = Maze(rows, cols)

    # Recursive backtracker
    visited = [[False] * cols for _ in range(rows)]
    stack = [(0, 0)]
    visited[0][0] = True
    while stack:
        r, c = stack[-1]
        neighbors = [
            (r + dr, c + dc)
            for d in DIRS
            for dr, dc in [DELTA[d]]
            if m.in_bounds(r + dr, c + dc) and not visited[r + dr][c + dc]
        ]
        if not neighbors:
            stack.pop()
        else:
            nr, nc = random.choice(neighbors)
            m.remove_wall(r, c, nr, nc)
            visited[nr][nc] = True
            stack.append((nr, nc))

    # Extra openings — Fisher-Yates puis on prend les N premiers
    walls = []
    for r in range(rows):
        for c in range(cols):
            if c + 1 < cols and m.has_wall(r, c, EAST):
                walls.append((r, c, r, c + 1))
            if r + 1 < rows and m.has_wall(r, c, SOUTH):
                walls.append((r, c, r + 1, c))
    random.shuffle(walls)
    for r1, c1, r2, c2 in walls[:extra_openings]:
        m.remove_wall(r1, c1, r2, c2)

    return m
